### ISSUE TITLE
[Feature Request]: `SimpleDirectoryReader` reading from AWS S3

### ISSUE THREAD

**Issue by jamesbraza** [run-llama/llama_index#8210]
### Feature Description

https://docs.llamaindex.ai/en/stable/core_modules/data_modules/storage/save_load.html#using-a-remote-backend

This demo has `SimpleDirectoryReader` reading in documents from local storage.  It would be cool if `SimpleDirectoryReader` could directly integrate with AWS S3 via `s3fs.S3FileSystem`.

### Reason

Not having to store raw corpus on local filesystem.

### Value of Feature

Increased flexibility when creating `VectorStore`s

**Comment by Disiok** [comment]
We have https://llamahub.ai/l/s3. Is this basically what the suggestion is?

**Comment by jamesbraza** [comment]
Thanks for sharing that link. A line from https://llamahub.ai/l/s3:

> All files are temporarily downloaded locally and subsequently parsed with `SimpleDirectoryReader`.

To me, this means `S3Reader` isn't "reading" from S3, it's downloading from S3 and then locally `open`ing the files. I can easily reinvent this with `boto3` in my own Python code. This request is about eliminating the local filesystem extra stop.

Ideally, `SimpleDirectoryReader` (or a true `S3Reader`) can directly read from AWS S3 into a Python process using something like the abstract filesystem specs.

**Comment by MrGobert** [comment]
I support this as I am trying to build my own logic that allows me to use PDF files from a cloud source as well (Google Cloud bucket) but I'm having issues when passing those processed files (Strings) into GPTSimpleVectorIndex. I'm getting the error: Error occurred while constructing index: Invalid document type: <class 'str'>.

Not sure what type the files have to be in order to be accepted? JSON? 
Directly supporting this would be awesome and save me some time. The goal is to be able to create a mobile version of my app and the first step is removing the need for locally stored documents to use with queries.

**Comment by nerdai** [comment]
Thanks @jamesbraza! I believe you're referring to using something like [s3fs](https://s3fs.readthedocs.io/en/latest/), correct?

**Comment by jamesbraza** [comment]
Yeah that's correct @nerdai 👌 feel free to ask for other clarifications if needed

**Comment by jamesbraza** [comment]
@MrGobert thanks for sharing, I agree with you it's useful to avoid the local filesystem entirely, in some use cases.

It sounds like your "Invalid document type: <class 'str'>." thing might be a separate issue, would you mind opening a new issue and providing the full stack trace there?  By GPTSimpleVectorIndex, did you actually mean `SimpleVectorStore`?

**Comment by utility-aagrawal** [comment]
A couple of other reasons this would be useful:

1) s3 loader downloads files in a temp directory locally and doesn't keep a track of file names. This makes it difficult to track them if you need them later to present sources of your RAG output. I opened an issue here before: https://github.com/run-llama/llama-hub/issues/611

2) You can't use document refresh() with this s3 loader since file names are generated randomly and refresh() uses filename_as_id flag.

Let me know in case of any questions. Thanks!

**Comment by rtjohn** [comment]
I support this as well for all the excellent reasons articulated already.

**Comment by MacOS** [comment]
Does anyone work on this? If not, I would be happy to help.

**Comment by nerdai** [comment]
> Does anyone work on this? If not, I would be happy to help, i.e. you can assign that to me if no one has already worked on it.

Thanks @MacOS for taking this on! With our v0.10 release, I am wondering if this should live as it own integration package. Please take a look at our [Contributing Guide](https://github.com/run-llama/llama_index/blob/main/CONTRIBUTING.md#creating-a-new-integration-package). I believe this could exist as a new reader integration.

**Comment by MacOS** [comment]
@nerdai Thank you for the pointers! I will keep you posted on my progress.

**Comment by sourabhdesai** [comment]
Hi @MacOS I have a PR here that adds this very feature!
https://github.com/run-llama/llama_index/pull/11303

It adds an optional `fs` parameter to `SimpleDirectoryReader` which can be any `fsspec` file-system object, including [`s3fs`](https://s3fs.readthedocs.io/en/latest/index.html).

Should be merged soon :)

**Comment by sourabhdesai** [comment]
@jamesbraza  @MacOS FYI [my PR](https://github.com/run-llama/llama_index/pull/11303) has been merged and should be available in the next release! feel free to re-open if there are any further concerns.

**Comment by MacOS** [comment]
@sourabhdesai Well, that was fast.

No concerns from my side.

**Comment by frenkd** [comment]
Hi! I'm using this on a project and it works great with .txt and .md files reading from S3. I get an error with .doc and .pdf files however:

`Failed to load file s3-bucket-name/media/documents/FILENAME.pdf with error: 'S3File' object has no attribute 'name'. Skipping...`

Any ideas on why this is happening only with certain filetypes?

**Comment by derinben** [comment]
[sourabhdesai](https://github.com/sourabhdesai) Thanks a lot for the addition of fsspec 👍 

I get the same error

`Failed to load file s3bucket/.../.../doc.pdf with error: 'S3File' object has no attribute 'name'. Skipping.` 

I remember encountering the same error a week ago but resolved it when updating llama-index but that doesn't seem to do the job now.

**Comment by thiagosalvatore** [comment]
Same thing happening to me when reading files

`with error: 'S3File' object has no attribute 'name'. Skipping.`
